### PR TITLE
feat(GC): tail upd

### PR DESCRIPTION
**Pull request by Kouprin**
Fixes https://github.com/nearprotocol/nearcore/issues/2413.
Fixes https://github.com/nearprotocol/nearcore/issues/2399.

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -540,27 +542,29 @@ impl Chain {
...
     // ===
     //
     // Prerequisites, guaranteed by the System:
-    // 1. Genesis is always on the Canonical Chain, only one Genesis exists and its height is `genesis_height`,
-    //    and no block with height lower than `genesis_height` exists.
-    // 2. If block A is ancestor of block B, height of A is strictly less then height of B.
-    // 3. (Property 1). An oldest block where fork is happened has the least height among all blocks on the fork.
-    // 4. (Property 2). An oldest block where fork is happened is never affected by Canonical Chain Switching
+    // 1. Genesis block is available and should not be removed by GC.
+    // 2. There is known lowest block height (Tail) which is the most earliest existing block came from
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/chain.rs:551]: Removed.

File: chain/chain/src/chain.rs [hunk 2]
```diff
@@ -540,27 +542,29 @@ impl Chain {
...
+    //    Genesis or State Sync.
+    // 3. Tail is always on the Canonical Chain, only one Tail exists and its height is higher than `genesis_height`,
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/chain.rs:553]: ++

File: chain/chain/src/chain.rs [hunk 3]
```diff
@@ -540,27 +542,29 @@ impl Chain {
...
+    //    and no block except Genesis has height lower or equal to `genesis_height`.
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/chain.rs:554]: ++

File: chain/chain/src/chain.rs [hunk 4]
```diff
@@ -540,27 +542,29 @@ impl Chain {
...
+    // 4. If block A is ancestor of block B, height of A is strictly less then height of B.
+    // 5. (Property 1). An oldest block where fork is happened has the least height among all blocks on the fork.
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/chain.rs:556]: Awesome.

[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/chain.rs:556]: A->B, B->C, B->D, D->E, D->F
there are forks from B and D, both are on the same fork
however, B is the oldest one and B has the least height among all blocks on the fork

File: chain/chain/src/chain.rs [hunk 5]
```diff
@@ -540,27 +542,29 @@ impl Chain {
...
-    // 2. `clear_data()` runs GC process for all blocks from the lowest known block height (tail).
-    // 3. `clear_data()` contains of four parts:
-    //    a. Block Reference Map creating, if it not exists
-    //    b. Define clearing height as `height`; the highest height for which we want to clear the data
-    //    c. Forks Clearing at height `height` down to tail.
-    //    d. Canonical Chain Clearing from tail up to height `height`
+    // 2. `clear_data()` runs GC process for all blocks from the Tail to GC Stop Height provided by Epoch Manager.
+    // 3. `clear_data()` executes separately:
+    //    a. Forks Clearing runs for each height from Tail up to GC Stop Height.
+    //    b. Canonical Chain Clearing from Tail up to GC Stop Height.
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/chain.rs:565]: ++

File: chain/chain/src/chain.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bowenwang1996) [reply chain/chain/src/chain.rs]: It seems to me that `A` is the block with the smallest height on that fork.

[the remark this answers is not in the document]
Comment (SkidanovAlex) [reply chain/chain/src/chain.rs]: I personally understand what KPR is saying ("An oldest block where fork is happened" -> "An oldest block among all the block at which the fork has happened" -> "An oldest block among all the blocks with the outgoing degree of 2 or more" -> "A block with the lowest height among all the blocks with outgoing degree of 2 or more").

We can replace with any desired level of specificity :)

[the remark this answers is not in the document]
Comment (Kouprin) [reply chain/chain/src/chain.rs]: ++

File: chain/chain/src/chain.rs [hunk 6]
```diff
@@ -544,23 +546,32 @@ impl Chain {
...
-    // 2. If block A is ancestor of block B, height of A is strictly less then height of B.
-    // 3. (Property 1). An oldest block where fork is happened has the least height among all blocks on the fork.
-    // 4. (Property 2). An oldest block where fork is happened is never affected by Canonical Chain Switching
+    // 1. Genesis block is available and should not be removed by GC.
+    // 2. No block in storage except Genesis has height lower or equal to `genesis_height`.
+    // 3. There is known lowest block height (Tail) came from Genesis or State Sync.
+    //    a. Tail is always on the Canonical Chain.
+    //    b. Only one Tail exists.
+    //    c. Tail's height is higher than or equal to `genesis_height`,
+    // 4. There is known highest block height (Head).
```
Comment (bowenwang1996) [inline chain/chain/src/chain.rs:555]:
```suggestion
    // 4. There is a known highest block height (Head).
```

File: chain/chain/src/chain.rs [hunk 7]
```diff
@@ -544,23 +546,32 @@ impl Chain {
...
+    //    a. Head is always on the Canonical Chain.
+    // 5. All blocks in the storage have heights in range [Tail; Head].
+    //    a. All forks end up on height of Head or lower.
+    // 6. If block A is ancestor of block B, height of A is strictly less then height of B.
+    // 7. (Property 1). A block with the lowest height among all the blocks at which the fork has started,
+    //    i.e. all the blocks with the outgoing degree 2 or more,
+    //    has the least height among all blocks on the fork.
+    // 8. (Property 2). An oldest block where fork is happened is never affected by Canonical Chain Switching
```
Comment (bowenwang1996) [inline chain/chain/src/chain.rs:563]: Can we at least fix the grammar here? "The oldest block where the fork happened..."

Comment (Kouprin) [reply chain/chain/src/chain.rs:563]: @bowenwang1996 please fix my grammar as much as possible :)

File: chain/chain/src/chain.rs (hunk 6 above)
Comment (Kouprin) [reply chain/chain/src/chain.rs:555]: ++

**Pull request merged**